flang/lib/Lower/ConvertVariable.cpp

```diff
@@ -335,6 +335,14 @@ static void instantiateGlobal(Fortran::lower::AbstractConverter &converter,
                                               global.getSymbol());
   Fortran::lower::StatementContext stmtCtx;
   mapSymbolAttributes(converter, var, symMap, stmtCtx, addrOf);
+  if (sym.test(Fortran::semantics::Symbol::Flag::InNamelist)) {
```

Comment:
The motivation for this alias code is to avoid duplicate `fir.address_of` ops.  This test shows the issue:

```
    module mmm
      namelist /nnn/ jjj
    end module

      use mmm
      jjj = 17
      write(*, nnn)
    end
```

In function `main` there are two symbols for `jjj`.  One is a direct local symbol, which is an alias for the other ultimate `jjj` symbol from the module.  The purpose of the `instantiateGlobal` code is to map both of these symbols to the same `fir.address` op.  Without this `instantiateGlobal` code, the generated high level FIR is:

```
func @_QQmain() {
  %0 = fir.alloca !fir.box<i32> {uniq_name = ""}
  %1 = fir.address_of(@_QMmmmEjjj) : !fir.ref<i32>
  %2 = fir.address_of(@_QMmmmEjjj) : !fir.ref<i32>
  %c17_i32 = constant 17 : i32
  fir.store %c17_i32 to %1 : !fir.ref<i32>
  . . .
  %11 = fir.embox %2 : (!fir.ref<i32>) -> !fir.box<i32>
```

Some optimization pass collapses these two `fir.address_of` ops into a single instance.  The `instantiateGlobal` code avoids this duplication up front.

For a namelist group defined in a module, it is probably safe to let optimization clean up this duplication because the namelist object is a global.  Another possibility though is an uplevel access to a local namelist
group object in a containing subprogram where the code would otherwise incorrectly generate addresses for two distinct objects, rather than duplicate addresses for a single object.  Since we don't yet have uplevel referencing, I'm not sure if this is actually a correctness issue or not.

Reply:
A little more thought suggests that this `instantiateGlobal` instance of the code is probably optional, but there might need to also be similar code in `instantiateLocal`.

Reply:
I do not think the same situation can happen in `instantiateLocal`, but I guess similar thing can happen with common/equivalenced. However, the consequences of this is only a double instantiation (once for the HostAssociated symbol, and once for the ultimate one) that CSE can indeed easily re-group as you mentioned. I would tend to not add special handling/complexity in lowering (from a pedantic way, your code still duplicates the specification expression evaluation, which are also simple expressions that gets CSEed).

If we really wants to take it at some point, I would rather that we find a way to re-group the HostAssociated/Ultimate symbols into a single variables during the PFT analysis rather than to have ad-hoc handling. 

Also, this case only happen with Namelist defined in modules, and you will most likely change the way this case is handled soon anyway (making the NamelistGroup a global), so this symbol will not be used anymore.

Reply:
Ok, I'll remove this code.  We can revisit this if we eventually find cases where having two different symbols for a namelist object causes generation of two objects, rather than just two object addresses.